server_environment/models/server_env_tech_name_mixin.py

```diff
@@ -39,6 +37,15 @@ class ServerEnvTechNameMixin(models.AbstractModel):
 
     _server_env_section_name_field = "tech_name"
 
+    @api.constrains('tech_name')
+    def _check_tech_name(self):
+
+        for record in self:
+            if hasattr(record, 'company_id') and self.search([('id', '!=', record.id), ('tech_name', '=' , record.tech_name), ('company_id', '=', record.company_id.id)]):
+                raise ValidationError(_("Tech name %s is duplicated. Tech name must be unique per company.") % record.tech_name)
+            if not hasattr(record, 'company_id') and self.search([('id', '!=', record.id), ('tech_name', '=' , record.tech_name)]):
+                raise ValidationError(_("Tech name %s is duplicated. Tech name must be unique.") % record.tech_name)
+                
```

Comment:
If you want to keep your proposal please adjust it to be similar to this

```suggestion
    _tech_name_unique_per_company = False   # off by default
    [...]
    @api.constrains("tech_name", "company_id")
    def _check_tech_name(self):
        company_dependent = self._tech_name_unique_per_company and  "company_id" in self._fields
        for record in self:
            domain = [('id', '!=', record.id), ('tech_name', '=' , record.tech_name)]
            error_msg = _("Tech name %s is duplicated. Tech name must be unique") % record.tech_name
            if company_dependent:
                domain.append(('company_id', '=', record.company_id.id))
                error_msg += _(" per company")
            if self.search_count(domain):
                raise ValidationError(error_msg)
```
and then add a test :pray:

Reply:
Good idea :+1:  I applied that code

Reply:
I'm afraid this piece of code is not enough. Existing installation will have the constraint in the DB.
We need an upgrade step that looks into all inheriting models and drops the constraint.
Furthermore, I'd like to see some tests.
If this is too much for you/your project right now, I would add a TODO w/ a reference to this PR and stick to a proper computation of the name using the company ID or a special field.

Reply:
I dropped the constraint in the project where I am using that code. I'll spend some time to apply this improvements